### PR TITLE
[tier1-cloud] Add Terraform section

### PR DESCRIPTION
**Pull request by alvarosimon**

### CODE REVIEW

File: intro-Cloud/ch_Terraform.tex [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+HashiCorp \gls{Terraform} \url{https://www.terraform.io/} is an infrastructure
+as code tool (IaC), similar to OpenStack \gls{Heat} orchestrator (\ref{cha:orch-using-heat}).
+Users can deploy a data center infrastructure using a declarative
+configuration language known as HashiCorp Configuration Language (HCL), or using JSON.
+\gls{Terraform} has some advantages over OpenStack \gls{Heat} service.
+It is has a simple syntax, it can provision virtual infrastructures across multiple cloud
+providers (not only OpenStack) and it provides important features not supported
+by \gls{Heat} at this moment, like network port forwarding rules (see \ref{sec:floating-ip}).
+This means with Terraform scripts like \lstinline{neutron_port_forward}
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:11]: that with Terraform

File: intro-Cloud/ch_Terraform.tex [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+(see \ref{port-forwarding}) are not longer needed.
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:12]: no longer needed

File: intro-Cloud/ch_Terraform.tex [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        
+    region_name: "regionOne"
+        
+      
+    interface: "public"
+    identity_api_version: 3
+    auth_type: "v3applicationcredential"
+\end{code}
+
+As the file comments state, you should copy the current clouds.yaml to your VSC login node \$home
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:63]: $home -> $HOME ?

File: intro-Cloud/ch_Terraform.tex [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+Make sure you have \texttt{\emph{\~\//.config/openstack/clouds.yaml}} available from the login node (see previous section \ref{sec:app-cred-terraform}).
+
+\strong{Note:} Do not share your application credential file clouds.yaml or put this file in a public place.
+
+\begin{prompt}
+%\shellcmd{chmod 600 ~/.config/openstack/clouds.yaml}
+\end{prompt}
+
+\section{Generate Terraform template variables}\label{sec:generate-terraform-variables}
+\gls{Terraform} requires some variables to know which resources are avaiable from the cloud provider for the user or project.
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:94]: available

File: intro-Cloud/ch_Terraform.tex [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+You do not have to include these variables manually, we have included a script to gather these variable IDs automatically.
+From Terraform directory cloned from git in the previous step (see: \ref{sec:getting-terraform-templ}), got to scripts directory:
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:96]: From the Terraform directory cloned from git in the previous step (see: \ref{sec:getting-terraform-templ}), go to the scripts directory:

File: intro-Cloud/ch_Terraform.tex [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+%\shellcmd{cd ~/openstack-templates/terraform/scripts}
+\end{prompt}
+
+And now run the script (usually you only have to run this script once).
+
+\begin{prompt}
+%\shellcmd{./modify\_variable.sh}
+\end{prompt}
+
+This step will take some seconds. The script will contact to VSC OpenStack public API to gather all the resources available for your project and fetch all the resource's IDs.
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:108]: will contact the VSC OpenStack public API

File: intro-Cloud/ch_Terraform.tex [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+(myvm)           ssh -p 56469 <user>@193.190.80.3
+(myvm-nginx)     ssh -p 59112 <user>@193.190.80.3
+(myvm-vsc_net)   ssh -p 54872 <user>@193.190.80.3
+(myvm-nfs_share) ssh -p 51280 <user>@193.190.80.3
+\end{prompt}
+
+After this step your Terraform templates will be ready to be deployed.
+
+\section{Modify default Terraform modules}\label{sec:modify-terraform-modules}
+In section \ref{sec:getting-terraform-templ} we have downloaded the Terraform module examples from VSC repository.
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:142]: from the VSC repository

File: intro-Cloud/ch_Terraform.tex [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+If you deploy these modules as it is it will deploy several VM examples by default such as:
+
+
+\begin{enumerate}
+\item \strong{myvm:} simple VM with 20Gb persistent volume and ssh access with port forwarding.
+\item \strong{myvm-nginx:} Like previous example but with an ansible playbook to install nginx and acces to port 80 besides ssh.
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:148]: access

File: intro-Cloud/ch_Terraform.tex [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+\item \strong{myvm-vsc\_net:} Similar to the fist example but also includes a VSC network interface (only available for some projects).
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:149]: first example

File: intro-Cloud/ch_Terraform.tex [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+..
+module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01: Still creating... [1m30s elapsed]
+module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01: Creation complete after 1m35s [id=88c7d037-5c44-45b7-acce-f5e4e58b1c35]
+
+Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
+\end{prompt}
+
+Your cloud infrastrucuture is ready to be used.
+
+It is important to keep a backup of your terraform directory, specially files within environment directory:
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:242]: especially all files within the environment directory

File: intro-Cloud/ch_Terraform.tex [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+\begin{prompt}
+%\shellcmd{~/openstack-templates/terraform/environment}
+\end{prompt}
+
+Terraform generates several files in this directory to keep track of any change in your infrastructure. If for some reason you lost or remove these files you will not able to modify or change the current Terraform plan (only directly from OpenStack).
+
+You can also modify and add more resources for the current templates. This task is out of the scope of this document, please refer to official Terraform documentation to add you own changes \url{https://www.terraform.io/docs} or ask to VSC Cloud admins via email at \cloudinfo.
+
```
Comment (epauwels) [inline intro-Cloud/ch_Terraform.tex:250]: I successfully followed the entire manual.
A first run of 'terraform apply' yielded this error for me though:

Error: Error waiting for instance (8f1960c8-2292-4c87-bc88-662883246fd4) to become ready: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)
│ 
│   with module.vm_with_pf_rules_with_ssh_access.openstack_compute_instance_v2.instance_01,
│   on ../modules/vm_with_pf_rules_with_ssh_access/main.tf line 1, in resource "openstack_compute_instance_v2" "instance_01":
│    1: resource "openstack_compute_instance_v2" "instance_01" {
│ 
╵
╷
│ Error: Error waiting for instance (8834a023-4068-4ef6-851e-08a05fe1392a) to become ready: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)
│ 
│   with module.vm_with_pf_rules_with_ssh_access_with_nginx.openstack_compute_instance_v2.instance_01,
│   on ../modules/vm_with_pf_rules_with_ssh_access_with_nginx/main.tf line 8, in resource "openstack_compute_instance_v2" "instance_01":
│    8: resource "openstack_compute_instance_v2" "instance_01" {
│ 

Second attempt yielded 'Apply complete'

**Comment by alvarosimon** [comment]
@epauwels thanks a lot for the review and testing. Remarks addressed.

**Pull request merged**